// bangla-tarikh/src/index.ts
import { Utils } from './utils';

module BanglaTarikh {
	/**
	 * Returns the year of the given date in number format
	 * @param date given date object (optional)
	 */
	export function year(date?: Date): number {
		const givenDate: Date = date || new Date();
		let noboborshoThisYear: Date = Utils.getNoboborshoDate(givenDate.getFullYear());

		const yearInEnglish = givenDate.getFullYear();
		let yearInBangla = 963 + (yearInEnglish - 1556);

		if (givenDate < noboborshoThisYear) {
			yearInBangla--;
		}

		return yearInBangla;
	}

	/**
 * Returns the year of the given date in bn language in string format
 * @param date given date object (optional)
 */
	export function yearInBangla(date?: Date): string {
		const yearValue: number = year(date);
		const yearInBangla = Utils.getBanglaDigit(yearValue);
		return yearInBangla;
	}

	/**
	 * Returns the month
	 * @param date Given Date
	 */
	export function month(date?: Date): number {
		const givenDate: Date = date || new Date();
		const diff = Utils.getDateDifferenceWithNoboborsho(givenDate);
		const isLeapYear: boolean = Utils.isLeapYear(givenDate.getFullYear());
		let month = 0;

		const Days = Utils.DaysOfMonth;

		if ((diff < -(Days.Chaitra + Days.Falgun + Days.Magh) && !isLeapYear) || (diff < -(Days.Chaitra + Days.FalgunLeapYear + Days.Magh) && isLeapYear)) {
			month = Utils.MonthOfYear.Poush;
		}
		else if ((diff < -(Days.Chaitra + Days.Falgun) && !isLeapYear) || (diff < -(Days.Chaitra + Days.FalgunLeapYear) && isLeapYear)) {
			month = Utils.MonthOfYear.Magh;
		}
		else if (diff < -Days.Chaitra) {
			month = Utils.MonthOfYear.Falgun;
		}
		else if (diff < 0) {
			month = Utils.MonthOfYear.Chaitra;
		}
		else if (diff <= Days.Baishakh - 1) {
			month = Utils.MonthOfYear.Baishakh;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha - 1) {
			month = Utils.MonthOfYear.Jaishtha;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar - 1) {
			month = Utils.MonthOfYear.Ashar;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban - 1) {
			month = Utils.MonthOfYear.Shraban;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra - 1) {
			month = Utils.MonthOfYear.Bhadra;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin - 1) {
			month = Utils.MonthOfYear.Ashwin;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin + Days.Kartikh - 1) {
			month = Utils.MonthOfYear.Kartikh;
		}
		else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin + Days.Kartikh + Days.Aghrahayan - 1) {
			month = Utils.MonthOfYear.Aghrahayan;
		}
		else {
			month = Utils.MonthOfYear.Poush;
		}

		return month;
	}

	/**
	 * Returns month in bangla
	 * @param date 
	 */
	export function monthInBangla(date?: Date): string {
		const banglaMonth = month(date);
		return Utils.getBanglaDigit(banglaMonth);
	}

	/**
	 * Returns month in bangla
	 * @param date 
	 */
	export function monthInBanglaString(date?: Date): string {
		const banglaMonth = month(date);
		return Utils.banglaMonthInBengali[banglaMonth];
	}

	/**
	 * Returns month in bangla
	 * @param date 
	 */
	export function day(date?: Date): number {
		const givenDate: Date = date || new Date();
		const diff: number = Utils.getDateDifferenceWithNoboborsho(givenDate);
		const banglaMonth: number = month(givenDate);

		let day: number = 1;

		const Days = Utils.DaysOfMonth;

		if (banglaMonth == Utils.MonthOfYear.Baishakh) {
			day = diff + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Jaishtha) {
			day = diff - Days.Jaishtha + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Ashar) {
			day = diff - Days.Jaishtha - Days.Ashar + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Shraban) {
			day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Bhadra) {
			day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Ashwin) {
			day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Kartikh) {
			day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Aghrahayan) {
			day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh - Days.Aghrahayan + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Poush) {
			if (diff > 0) {
				day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh - Days.Aghrahayan - Days.Poush + 1;
			} else {
				const isLeapYear: boolean = Utils.isLeapYear(givenDate.getFullYear());
				const falgun: number = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
				day = Days.Poush + Days.Magh + falgun + Days.Chaitra + diff + 1;
			}
		}
		else if (banglaMonth == Utils.MonthOfYear.Magh) {
			const isLeapYear: boolean = Utils.isLeapYear(givenDate.getFullYear());
			const falgun: number = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
			day = Days.Chaitra + Days.Magh + falgun + diff + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Falgun) {
			const isLeapYear: boolean = Utils.isLeapYear(givenDate.getFullYear());
			const falgun: number = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
			day = Days.Chaitra + falgun + diff + 1;
		}
		else if (banglaMonth == Utils.MonthOfYear.Chaitra) {
			day = Days.Chaitra + diff + 1;
		}

		return day;
	}

	export function dayInBangla(date?: Date): string {
		const givenDate: Date = date || new Date();
		const dayInBangla = day(givenDate);
		return Utils.getBanglaDigit(dayInBangla);
	}

	export function format(dateFormat?: string, date?: Date): string {
		const givenDate: Date = date || new Date();

		if (!dateFormat) {
			dateFormat = 'DTS MMMM, YYYYT';
		}

		let formattedDate: string = dateFormat.replace(/(DDD|DDT|DTS|DT|DD|D)/g, function (match: string) {
			const dayOfMonth: number = day(givenDate);

			if (match === 'DDD') {
				return Utils.numberToBanglaDetail[dayOfMonth];
			}
			else if (match === 'DDT') {
				const dayInString: string = dayOfMonth < 10 ? `0${dayOfMonth}` : dayOfMonth.toString();
				return Utils.getBanglaDigit(dayInString);
			}
			else if (match === 'DTS') {
				const dayInBangla: string = Utils.getBanglaDigit(dayOfMonth);
				return dayInBangla + Utils.daySuffix[dayOfMonth];
			}
			else if (match === 'DT') {
				return Utils.getBanglaDigit(dayOfMonth);
			}
			else if (match === 'DD') {
				return dayOfMonth < 10 ? `0${dayOfMonth}` : dayOfMonth.toString();
			}
			else {
				return dayOfMonth;
			}
		});

		formattedDate = formattedDate.replace(/(MMMM|MMT|MMM|MT|MM|M)/g, function (match: string) {
			const monthOfYear: number = month(givenDate);

			if (match === 'MMMM') {
				return Utils.banglaMonthInBengali[monthOfYear];
			}
			else if (match === 'MMT') {
				const monthInString: string = monthOfYear < 10 ? `0${monthOfYear}` : monthOfYear.toString();
				return Utils.getBanglaDigit(monthInString);
			}
			else if (match === 'MMM') {
				return Utils.banglaMonthInEnglish[monthOfYear];
			}
			else if (match === 'MT') {
				return Utils.getBanglaDigit(monthOfYear);
			}
			else if (match === 'MM') {
				return monthOfYear < 10 ? `0${monthOfYear}` : monthOfYear.toString();
			}
			else {
				return monthOfYear;
			}
		});

		formattedDate = formattedDate.replace(/(YYYYT|YYYY)/g, function (match: string) {
			const yearValue: number = year(givenDate);

			if (match === 'YYYYT') {
				return Utils.getBanglaDigit(yearValue);
			}
			else {
				return yearValue.toString();
			}
		});

		return formattedDate;
	}
}

export default BanglaTarikh;
