// bangla-tarikh/src/utils.ts
export module Utils {

	/**
	 * Bangla numbers
	 */
	export const banglaNumber: any = {
		'1': '১',
		'2': '২',
		'3': '৩',
		'4': '৪',
		'5': '৫',
		'6': '৬',
		'7': '৭',
		'8': '৮',
		'9': '৯',
		'0': '০'
	}

	/**
	 * Bangla months
	 */
	export const banglaMonthInBengali: any = {
		'1': 'বৈশাখ',
		'2': 'জ্যৈষ্ঠ',
		'3': 'আষাঢ়',
		'4': 'শ্রাবণ',
		'5': 'ভাদ্র',
		'6': 'আশ্বিন',
		'7': 'কার্তিক',
		'8': 'অগ্রহায়ণ',
		'9': 'পৌষ',
		'10': 'মাঘ',
		'11': 'ফাল্গুন',
		'12': 'চৈত্র',
	}

	export const banglaMonthInEnglish: any = {
		'1': 'Boishakh',
		'2': 'Jaishtha',
		'3': 'Ashar',
		'4': 'Shraban',
		'5': 'Bhadra',
		'6': 'Ashwin',
		'7': 'Kartik',
		'8': 'Aghrahayan',
		'9': 'Poush',
		'10': 'Magh',
		'11': 'Falgun',
		'12': 'Chaitra',
	}

	export enum MonthOfYear {
		Baishakh = 1,
		Jaishtha = 2,
		Ashar = 3,
		Shraban = 4,
		Bhadra = 5,
		Ashwin = 6,
		Kartikh = 7,
		Aghrahayan = 8,
		Poush = 9,
		Magh = 10,
		Falgun = 11,
		Chaitra = 12
	}

	export enum DaysOfMonth {
		Baishakh = 31,
		Jaishtha = 31,
		Ashar = 31,
		Shraban = 31,
		Bhadra = 31,
		Ashwin = 30,
		Kartikh = 30,
		Aghrahayan = 30,
		Poush = 30,
		Magh = 30,
		Falgun = 30,
		Chaitra = 30,
		FalgunLeapYear = 31
	}

	export const numberToBanglaDetail: any = {
		'1': 'পহেলা',
		'2': 'দোসরা',
		'3': 'তেসরা',
		'4': 'চৌঠা',
		'5': 'পাঁচই',
		'6': 'ছউই',
		'7': 'সাতই',
		'8': 'আটই',
		'9': 'নয়ই',
		'10': 'দশই',
		'11': 'এগারোই',
		'12': 'বারোই',
		'13': 'তেরোই',
		'14': 'চৌদ্দই',
		'15': 'পনেরোই',
		'16': 'ষোলোই',
		'17': 'সতেরোই',
		'18': 'আঠারোই',
		'19': 'উনিশে',
		'20': 'বিশই',
		'21': 'একুশে',
		'22': 'বাইশে',
		'23': 'তেইশে',
		'24': 'চব্বিশে',
		'25': 'পঁচিশে',
		'26': 'ছাব্বিশে',
		'27': 'সাতাশে',
		'28': 'আটাশে',
		'29': 'ঊনত্রিশে',
		'30': 'ত্রিশে',
		'31': 'একত্রিশে'
	}

	export const daySuffix: any = {
		'1': 'লা',
		'2': 'রা',
		'3': 'রা',
		'4': 'ঠা',
		'5': 'ই',
		'6': 'ই',
		'7': 'ই',
		'8': 'ই',
		'9': 'ই',
		'10': 'ই',
		'11': 'ই',
		'12': 'ই',
		'13': 'ই',
		'14': 'ই',
		'15': 'ই',
		'16': 'ই',
		'17': 'ই',
		'18': 'ই',
		'19': 'শে',
		'20': 'শে',
		'21': 'শে',
		'22': 'শে',
		'23': 'শে',
		'24': 'শে',
		'25': 'শে',
		'26': 'শে',
		'27': 'শে',
		'28': 'শে',
		'29': 'শে',
		'30': 'শে',
		'31': 'শে'
	}

	export const numberOfDaysInBanglaCalender: Array<number> = [31, 31, 31, 31, 31, 30, 30, 30, 30, 30, 30, 30];

	/**
	 * Returns numeric value to equivalent bangla font in string format
	 * @param data numeric values in string format
	 */
	export function getBanglaDigit(data: string | number): string {
		return data.toString().replace(/\d/g, function (match: string) {
			return banglaNumber[match];
		});
	}

	/**
	 * Returns noboborsho date for given year
	 * @param year
	 */
	export function getNoboborshoDate(year: number): Date {
		return new Date(year, 3, 14, 0, 0, 0);
	}

	export function isLeapYear(year: number): boolean {
		return ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0);
	}

	/**
	 * Return the date difference of a date with noboborsho of that year
	 * @param date Given date object
	 */
	export function getDateDifferenceWithNoboborsho(date: Date): number {
		const givenDate: Date = new Date(date.getFullYear(), date.getMonth(), date.getDate());
		const noboborshoThisYear: Date = Utils.getNoboborshoDate(givenDate.getFullYear());
		const diff = Math.ceil((givenDate.getTime() - noboborshoThisYear.getTime()) / (24 * 3600 * 1000));
		return diff;
	}
}
