// bangla-tarikh/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("./utils");
var BanglaTarikh;
(function (BanglaTarikh) {
    /**
     * Returns the year of the given date in number format
     * @param date given date object (optional)
     */
    function year(date) {
        var givenDate = date || new Date();
        var noboborshoThisYear = utils_1.Utils.getNoboborshoDate(givenDate.getFullYear());
        var yearInEnglish = givenDate.getFullYear();
        var yearInBangla = 963 + (yearInEnglish - 1556);
        if (givenDate < noboborshoThisYear) {
            yearInBangla--;
        }
        return yearInBangla;
    }
    BanglaTarikh.year = year;
    /**
 * Returns the year of the given date in bn language in string format
 * @param date given date object (optional)
 */
    function yearInBangla(date) {
        var yearValue = year(date);
        var yearInBangla = utils_1.Utils.getBanglaDigit(yearValue);
        return yearInBangla;
    }
    BanglaTarikh.yearInBangla = yearInBangla;
    /**
     * Returns the month
     * @param date Given Date
     */
    function month(date) {
        var givenDate = date || new Date();
        var diff = utils_1.Utils.getDateDifferenceWithNoboborsho(givenDate);
        var isLeapYear = utils_1.Utils.isLeapYear(givenDate.getFullYear());
        var month = 0;
        var Days = utils_1.Utils.DaysOfMonth;
        if ((diff < -(Days.Chaitra + Days.Falgun + Days.Magh) && !isLeapYear) || (diff < -(Days.Chaitra + Days.FalgunLeapYear + Days.Magh) && isLeapYear)) {
            month = utils_1.Utils.MonthOfYear.Poush;
        }
        else if ((diff < -(Days.Chaitra + Days.Falgun) && !isLeapYear) || (diff < -(Days.Chaitra + Days.FalgunLeapYear) && isLeapYear)) {
            month = utils_1.Utils.MonthOfYear.Magh;
        }
        else if (diff < -Days.Chaitra) {
            month = utils_1.Utils.MonthOfYear.Falgun;
        }
        else if (diff < 0) {
            month = utils_1.Utils.MonthOfYear.Chaitra;
        }
        else if (diff <= Days.Baishakh - 1) {
            month = utils_1.Utils.MonthOfYear.Baishakh;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha - 1) {
            month = utils_1.Utils.MonthOfYear.Jaishtha;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar - 1) {
            month = utils_1.Utils.MonthOfYear.Ashar;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban - 1) {
            month = utils_1.Utils.MonthOfYear.Shraban;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra - 1) {
            month = utils_1.Utils.MonthOfYear.Bhadra;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin - 1) {
            month = utils_1.Utils.MonthOfYear.Ashwin;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin + Days.Kartikh - 1) {
            month = utils_1.Utils.MonthOfYear.Kartikh;
        }
        else if (diff <= Days.Baishakh + Days.Jaishtha + Days.Ashar + Days.Shraban + Days.Bhadra + Days.Ashwin + Days.Kartikh + Days.Aghrahayan - 1) {
            month = utils_1.Utils.MonthOfYear.Aghrahayan;
        }
        else {
            month = utils_1.Utils.MonthOfYear.Poush;
        }
        return month;
    }
    BanglaTarikh.month = month;
    /**
     * Returns month in bangla
     * @param date
     */
    function monthInBangla(date) {
        var banglaMonth = month(date);
        return utils_1.Utils.getBanglaDigit(banglaMonth);
    }
    BanglaTarikh.monthInBangla = monthInBangla;
    /**
     * Returns month in bangla
     * @param date
     */
    function monthInBanglaString(date) {
        var banglaMonth = month(date);
        return utils_1.Utils.banglaMonthInBengali[banglaMonth];
    }
    BanglaTarikh.monthInBanglaString = monthInBanglaString;
    /**
     * Returns month in bangla
     * @param date
     */
    function day(date) {
        var givenDate = date || new Date();
        var diff = utils_1.Utils.getDateDifferenceWithNoboborsho(givenDate);
        var banglaMonth = month(givenDate);
        var day = 1;
        var Days = utils_1.Utils.DaysOfMonth;
        if (banglaMonth == utils_1.Utils.MonthOfYear.Baishakh) {
            day = diff + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Jaishtha) {
            day = diff - Days.Jaishtha + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Ashar) {
            day = diff - Days.Jaishtha - Days.Ashar + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Shraban) {
            day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Bhadra) {
            day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Ashwin) {
            day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Kartikh) {
            day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Aghrahayan) {
            day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh - Days.Aghrahayan + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Poush) {
            if (diff > 0) {
                day = diff - Days.Jaishtha - Days.Ashar - Days.Shraban - Days.Bhadra - Days.Ashwin - Days.Kartikh - Days.Aghrahayan - Days.Poush + 1;
            }
            else {
                var isLeapYear = utils_1.Utils.isLeapYear(givenDate.getFullYear());
                var falgun = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
                day = Days.Poush + Days.Magh + falgun + Days.Chaitra + diff + 1;
            }
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Magh) {
            var isLeapYear = utils_1.Utils.isLeapYear(givenDate.getFullYear());
            var falgun = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
            day = Days.Chaitra + Days.Magh + falgun + diff + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Falgun) {
            var isLeapYear = utils_1.Utils.isLeapYear(givenDate.getFullYear());
            var falgun = isLeapYear ? Days.FalgunLeapYear : Days.Falgun;
            day = Days.Chaitra + falgun + diff + 1;
        }
        else if (banglaMonth == utils_1.Utils.MonthOfYear.Chaitra) {
            day = Days.Chaitra + diff + 1;
        }
        return day;
    }
    BanglaTarikh.day = day;
    function dayInBangla(date) {
        var givenDate = date || new Date();
        var dayInBangla = day(givenDate);
        return utils_1.Utils.getBanglaDigit(dayInBangla);
    }
    BanglaTarikh.dayInBangla = dayInBangla;
    function format(dateFormat, date) {
        var givenDate = date || new Date();
        if (!dateFormat) {
            dateFormat = 'DTS MMMM, YYYYT';
        }
        var formattedDate = dateFormat.replace(/(DDD|DDT|DTS|DT|DD|D)/g, function (match) {
            var dayOfMonth = day(givenDate);
            if (match === 'DDD') {
                return utils_1.Utils.numberToBanglaDetail[dayOfMonth];
            }
            else if (match === 'DDT') {
                var dayInString = dayOfMonth < 10 ? "0" + dayOfMonth : dayOfMonth.toString();
                return utils_1.Utils.getBanglaDigit(dayInString);
            }
            else if (match === 'DTS') {
                var dayInBangla_1 = utils_1.Utils.getBanglaDigit(dayOfMonth);
                return dayInBangla_1 + utils_1.Utils.daySuffix[dayOfMonth];
            }
            else if (match === 'DT') {
                return utils_1.Utils.getBanglaDigit(dayOfMonth);
            }
            else if (match === 'DD') {
                return dayOfMonth < 10 ? "0" + dayOfMonth : dayOfMonth.toString();
            }
            else {
                return dayOfMonth;
            }
        });
        formattedDate = formattedDate.replace(/(MMMM|MMT|MMM|MT|MM|M)/g, function (match) {
            var monthOfYear = month(givenDate);
            if (match === 'MMMM') {
                return utils_1.Utils.banglaMonthInBengali[monthOfYear];
            }
            else if (match === 'MMT') {
                var monthInString = monthOfYear < 10 ? "0" + monthOfYear : monthOfYear.toString();
                return utils_1.Utils.getBanglaDigit(monthInString);
            }
            else if (match === 'MMM') {
                return utils_1.Utils.banglaMonthInEnglish[monthOfYear];
            }
            else if (match === 'MT') {
                return utils_1.Utils.getBanglaDigit(monthOfYear);
            }
            else if (match === 'MM') {
                return monthOfYear < 10 ? "0" + monthOfYear : monthOfYear.toString();
            }
            else {
                return monthOfYear;
            }
        });
        formattedDate = formattedDate.replace(/(YYYYT|YYYY)/g, function (match) {
            var yearValue = year(givenDate);
            if (match === 'YYYYT') {
                return utils_1.Utils.getBanglaDigit(yearValue);
            }
            else {
                return yearValue.toString();
            }
        });
        return formattedDate;
    }
    BanglaTarikh.format = format;
})(BanglaTarikh || (BanglaTarikh = {}));
exports.default = BanglaTarikh;
