// bangla-tarikh/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils;
(function (Utils) {
    /**
     * Bangla numbers
     */
    Utils.banglaNumber = {
        '1': '১',
        '2': '২',
        '3': '৩',
        '4': '৪',
        '5': '৫',
        '6': '৬',
        '7': '৭',
        '8': '৮',
        '9': '৯',
        '0': '০'
    };
    /**
     * Bangla months
     */
    Utils.banglaMonthInBengali = {
        '1': 'বৈশাখ',
        '2': 'জ্যৈষ্ঠ',
        '3': 'আষাঢ়',
        '4': 'শ্রাবণ',
        '5': 'ভাদ্র',
        '6': 'আশ্বিন',
        '7': 'কার্তিক',
        '8': 'অগ্রহায়ণ',
        '9': 'পৌষ',
        '10': 'মাঘ',
        '11': 'ফাল্গুন',
        '12': 'চৈত্র',
    };
    Utils.banglaMonthInEnglish = {
        '1': 'Boishakh',
        '2': 'Jaishtha',
        '3': 'Ashar',
        '4': 'Shraban',
        '5': 'Bhadra',
        '6': 'Ashwin',
        '7': 'Kartik',
        '8': 'Aghrahayan',
        '9': 'Poush',
        '10': 'Magh',
        '11': 'Falgun',
        '12': 'Chaitra',
    };
    var MonthOfYear;
    (function (MonthOfYear) {
        MonthOfYear[MonthOfYear["Baishakh"] = 1] = "Baishakh";
        MonthOfYear[MonthOfYear["Jaishtha"] = 2] = "Jaishtha";
        MonthOfYear[MonthOfYear["Ashar"] = 3] = "Ashar";
        MonthOfYear[MonthOfYear["Shraban"] = 4] = "Shraban";
        MonthOfYear[MonthOfYear["Bhadra"] = 5] = "Bhadra";
        MonthOfYear[MonthOfYear["Ashwin"] = 6] = "Ashwin";
        MonthOfYear[MonthOfYear["Kartikh"] = 7] = "Kartikh";
        MonthOfYear[MonthOfYear["Aghrahayan"] = 8] = "Aghrahayan";
        MonthOfYear[MonthOfYear["Poush"] = 9] = "Poush";
        MonthOfYear[MonthOfYear["Magh"] = 10] = "Magh";
        MonthOfYear[MonthOfYear["Falgun"] = 11] = "Falgun";
        MonthOfYear[MonthOfYear["Chaitra"] = 12] = "Chaitra";
    })(MonthOfYear = Utils.MonthOfYear || (Utils.MonthOfYear = {}));
    var DaysOfMonth;
    (function (DaysOfMonth) {
        DaysOfMonth[DaysOfMonth["Baishakh"] = 31] = "Baishakh";
        DaysOfMonth[DaysOfMonth["Jaishtha"] = 31] = "Jaishtha";
        DaysOfMonth[DaysOfMonth["Ashar"] = 31] = "Ashar";
        DaysOfMonth[DaysOfMonth["Shraban"] = 31] = "Shraban";
        DaysOfMonth[DaysOfMonth["Bhadra"] = 31] = "Bhadra";
        DaysOfMonth[DaysOfMonth["Ashwin"] = 30] = "Ashwin";
        DaysOfMonth[DaysOfMonth["Kartikh"] = 30] = "Kartikh";
        DaysOfMonth[DaysOfMonth["Aghrahayan"] = 30] = "Aghrahayan";
        DaysOfMonth[DaysOfMonth["Poush"] = 30] = "Poush";
        DaysOfMonth[DaysOfMonth["Magh"] = 30] = "Magh";
        DaysOfMonth[DaysOfMonth["Falgun"] = 30] = "Falgun";
        DaysOfMonth[DaysOfMonth["Chaitra"] = 30] = "Chaitra";
        DaysOfMonth[DaysOfMonth["FalgunLeapYear"] = 31] = "FalgunLeapYear";
    })(DaysOfMonth = Utils.DaysOfMonth || (Utils.DaysOfMonth = {}));
    Utils.numberToBanglaDetail = {
        '1': 'পহেলা',
        '2': 'দোসরা',
        '3': 'তেসরা',
        '4': 'চৌঠা',
        '5': 'পাঁচই',
        '6': 'ছউই',
        '7': 'সাতই',
        '8': 'আটই',
        '9': 'নয়ই',
        '10': 'দশই',
        '11': 'এগারোই',
        '12': 'বারোই',
        '13': 'তেরোই',
        '14': 'চৌদ্দই',
        '15': 'পনেরোই',
        '16': 'ষোলোই',
        '17': 'সতেরোই',
        '18': 'আঠারোই',
        '19': 'উনিশে',
        '20': 'বিশই',
        '21': 'একুশে',
        '22': 'বাইশে',
        '23': 'তেইশে',
        '24': 'চব্বিশে',
        '25': 'পঁচিশে',
        '26': 'ছাব্বিশে',
        '27': 'সাতাশে',
        '28': 'আটাশে',
        '29': 'ঊনত্রিশে',
        '30': 'ত্রিশে',
        '31': 'একত্রিশে'
    };
    Utils.daySuffix = {
        '1': 'লা',
        '2': 'রা',
        '3': 'রা',
        '4': 'ঠা',
        '5': 'ই',
        '6': 'ই',
        '7': 'ই',
        '8': 'ই',
        '9': 'ই',
        '10': 'ই',
        '11': 'ই',
        '12': 'ই',
        '13': 'ই',
        '14': 'ই',
        '15': 'ই',
        '16': 'ই',
        '17': 'ই',
        '18': 'ই',
        '19': 'শে',
        '20': 'শে',
        '21': 'শে',
        '22': 'শে',
        '23': 'শে',
        '24': 'শে',
        '25': 'শে',
        '26': 'শে',
        '27': 'শে',
        '28': 'শে',
        '29': 'শে',
        '30': 'শে',
        '31': 'শে'
    };
    Utils.numberOfDaysInBanglaCalender = [31, 31, 31, 31, 31, 30, 30, 30, 30, 30, 30, 30];
    /**
     * Returns numeric value to equivalent bangla font in string format
     * @param data numeric values in string format
     */
    function getBanglaDigit(data) {
        return data.toString().replace(/\d/g, function (match) {
            return Utils.banglaNumber[match];
        });
    }
    Utils.getBanglaDigit = getBanglaDigit;
    /**
     * Returns noboborsho date for given year
     * @param year
     */
    function getNoboborshoDate(year) {
        return new Date(year, 3, 14, 0, 0, 0);
    }
    Utils.getNoboborshoDate = getNoboborshoDate;
    function isLeapYear(year) {
        return ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0);
    }
    Utils.isLeapYear = isLeapYear;
    /**
     * Return the date difference of a date with noboborsho of that year
     * @param date Given date object
     */
    function getDateDifferenceWithNoboborsho(date) {
        var givenDate = new Date(date.getFullYear(), date.getMonth(), date.getDate());
        var noboborshoThisYear = Utils.getNoboborshoDate(givenDate.getFullYear());
        var diff = Math.ceil((givenDate.getTime() - noboborshoThisYear.getTime()) / (24 * 3600 * 1000));
        return diff;
    }
    Utils.getDateDifferenceWithNoboborsho = getDateDifferenceWithNoboborsho;
})(Utils = exports.Utils || (exports.Utils = {}));
